File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        int i,j;
        k = k%n;
        //reverse last k elements
        for(i = n-k,j = n-1;i < j;i++,j--){
            swap(nums[i],nums[j]);
        } 
        
        //reverse first n-k elements
        for(i = 0,j = n-k-1;i < j;i++,j--){
            swap(nums[i],nums[j]);
        }

        //reverse the whole list
        for(i = 0,j = n-1;i <j;i++,j--){
            swap(nums[i],nums[j]);
        }
    }
};

File: LeetCode/Medium/0189-rotate-array/0189-rotate-array.java
class Solution {
    public void reverse(int[] arr,int start,int end){
        while(start < end){

            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }
    }
    public void rotate(int[] arr, int k) {
        int n = arr.length;

        k %= n;
        
        //reverse whole array
        reverse(arr,0,n-1);

        //reverse first k element
        reverse(arr,0,k-1);

        //reverse last k element
        reverse(arr,k,n-1);

        
        
    }
}